crates/server/src/authority/catalog.rs

```diff
@@ -92,6 +95,29 @@ impl RequestHandler for Catalog {
                 };
             }
 
+            // RFC 5001 "DNS Name Server Identifier (NSID) Option" handling.
+            match (req_edns.option(EdnsCode::NSID), &self.nsid_payload) {
+                // "A name server MUST NOT send an NSID option back to a resolver which
+                // did not request it."
+                (None, ..) => {}
+                // NSID was requested, but we don't have a payload configured.
+                (Some(_), None) => {
+                    trace!("ignoring EDNS NSID request - no response payload configured")
+                }
+                // NSID was requested, and we have a payload to reply with. Add it to the
+                // response EDNS.
+                (Some(request_option), Some(payload)) => {
+                    // "The name server MUST ignore any NSID payload data that might be
+                    //  present in the query message."
+                    if !request_option.is_empty() {
+                        warn!("ignoring non-empty EDNS NSID request payload")
+                    }
+                    resp_edns
+                        .options_mut()
+                        .insert(EdnsOption::NSID(payload.clone()));
```

Comment:
I have a slight preference for the `(Some(_), Some(_))` arm coming first, since it's the most interesting. Given that the order doesn't matter here, would sort the `(None, _)` arm last.

Reply:
Fixed.